### ISSUE TITLE
Support named groups in match

### ISSUE THREAD

**Issue by AtnNn** [rethinkdb/rethinkdb#3761]
RE2 regular expressions support named groups, but these names are ignored by RethinkDB. currently `r.expr("foobar").match("(?P<name>b.)(.)")` returns

```
{
    "end": 5 ,
    "groups": [
        {
            "end": 5 ,
            "start": 3 ,
            "str": "ba"
        },
        {
            "end": 6 ,
            "start": 5 ,
            "str": "r"
        } 
    ] ,
    "start": 3 ,
    "str": "ba"
}
```

It would be nice if the `groups` property was an object instead:

```
    "groups": {
        "name": {
            "end": 5 ,
            "start": 3 ,
            "str": "ba"
        },
        "1": {
            "end": 6 ,
            "start": 5 ,
            "str": "r"
        } 
    }
```

This would make dealing with large regular expressions a lot easier.

**Comment by mlucy** [comment]
Would we make `groups` an object even if there are only unnamed groups?  What should the output be if there are a mix of named and unnamed groups and one of the group names happens to be a number that matches the offset of one of the unnamed groups?

**Comment by danielmewes** [comment]
If we want to stay compatible with the current format, we could add the name as a property to the groups instead:
```
{
    "end": 5 ,
    "groups": [
        {
            "name": "name" ,
            "end": 5 ,
            "start": 3 ,
            "str": "ba"
        },
        {
            "end": 6 ,
            "start": 5 ,
            "str": "r"
        } 
    ] ,
    "start": 3 ,
    "str": "ba"
}
```

It's not very nice to use though if you want to access a specific named group directly.

I think it would be better to break compatibility here. It's pretty easy to get the old format back by using the discussed `values()` term (https://github.com/rethinkdb/rethinkdb/issues/2945).

In @AtnNn's proposal, unnamed groups are given numbers as keys. This sounds like a reasonable thing to do. I think we should do that even if there are no named groups at all for consistency.

**Comment by mlucy** [comment]
@danielmewes -- my question about using numbers as keys is "what if someone names one of his groups the same number as one of the offsets of an unnamed group"?  I guess we could just detect that and error.

I guess I'm OK with breaking backward compatibility here; the interface does seem nicer.

**Comment by danielmewes** [comment]
> "what if someone names one of his groups the same number as one of the offsets of an unnamed group"? I guess we could just detect that and error.

That sounds reasonable.

**Comment by mlucy** [comment]
Marking as settled; we'll always use the new format (so we'll break backward compatibility) and we'll error if numeric offsets conflict with the name of a group.

**Comment by marshall007** [comment]
One slightly annoying/potentially confusing thing to note concerning the new format:

```js
r.expr('foo').match('(f)')('groups')(0) // RqlRuntimeError: Cannot convert OBJECT to SEQUENCE
r.expr('foo').match('(f)')('groups')('0')
```

Perhaps when you use the "bracket" command on an object it would make sense to coerce the value to a string instead of assuming you want to treat it as a sequence? That way the two examples above would be equivalent.

**Comment by danielmewes** [comment]
@marshall007 Very good point, we had talked about this today as well actually. I think your proposal is reasonable, and we should consider doing that before merging the `match` change.

**Comment by mlucy** [comment]
@danielmewes -- I think automatically converting object keys to strings deserves its own ReQL discussion.

**Comment by danielmewes** [comment]
Yeah was just about to open one: https://github.com/rethinkdb/rethinkdb/issues/4140

**Comment by danielmewes** [comment]
Moving to subsequent, since #4140 would make this work without breaking backwards compatiblity for certain types of queries, which would be neat (it would still break backwards compatiblity for queries that don't use the bracket term on the `group` field).
